src/DIContainerTrait.php

```diff
@@ -88,4 +88,94 @@ protected function setMissingProperty($key, $value)
             ->addMoreInfo('property', $key)
             ->addMoreInfo('value', $value);
     }
+
+    /**
+     * Return the argument and assert it is instance of current class.
+     *
+     * The best, typehinting-friendly, way to annotate object type if it not defined
+     * at method header or strong typing in method header can not be used.
+     *
+     * @return static
+     */
+    public static function assertInstanceOf(object $object)// :static supported by PHP8+
+    {
+        if (!($object instanceof static)) {
+            throw (new Exception('Object is not an instance of static class'))
+                ->addMoreInfo('static_class', static::class)
+                ->addMoreInfo('object_class', get_class($object));
+        }
+
+        return $object;
+    }
+
+    private static function _fromSeedPrecheck($seed, bool $unsafe)// :self is too strict with unsafe behaviour
+    {
+        if (!is_object($seed)) {
+            if (!is_array($seed)) {
+                if (!is_string($seed)) { // allow string class name seed but prevent bad usage
+                    throw (new Exception('Seed must be an array, a string class name (deprecated) or an object'))
+                        ->addMoreInfo('seed_type', gettype($seed));
+                }
+
+                $seed = [$seed];
+            }
+
+            if (!isset($seed[0])) {
+                throw (new Exception('Class name is not specified by the seed'))
+                    ->addMoreInfo('seed', $seed);
+            }
+
+            $cl = $seed[0];
+            if (!$unsafe && !is_a($cl, static::class, true)) {
+                throw (new Exception('Seed class is not a subtype of static class'))
+                    ->addMoreInfo('static_class', static::class)
+                    ->addMoreInfo('seed_class', $cl);
+            }
+        }
+
+        return $seed;
+    }
+
+    /**
+     * Create new object from seed and assert it is instance of current class.
+     *
+     * The best, typehinting-friendly, way to create an object if it should not be
+     * immediately added to a parent (otherwise use addTo() method).
+     *
+     * @param array|string|object $seed the first element specifies a class name, other elements are seed
+     *
+     * @return static
+     */
+    public static function fromSeed($seed = [], $defaults = [])// :static supported by PHP8+
+    {
+        if (func_num_args() > 2) { // prevent bad usage
```

Comment:
You can reduce code here by simpy calling
```
$object = static::fromSeedUnsafe($seed, $defaults);

static::assertInstanceOf($object);

return $object;
```

Reply:
⚠️ Can, but do not want - this way, `assertInstanceOf` is computed even before the object is even created, which implies stronger security! (as seed can be an user input)

Reply:
What I see the `assertInstanceOf` is called before the `$object` is returned only?
The rest of the code is the same.

Reply:
⚠️ No! `_fromSeedPrecheck` is called BEFORE object is constructed. Be very carefull around this and test for this behaviour will be highly welcomed.

Reply:
PR for test: https://github.com/atk4/core/pull/311